engine/resource/src/resource_mounts.cpp

```diff
@@ -0,0 +1,539 @@
+    {
+        MountFileEntry& entry = entries[i];
+
+        DM_RESOURCE_DBG_LOG(2, "  mounting: '%s' %d '%s'\n", entry.m_Name, entry.m_Priority, entry.m_Uri);
+        LoadMount(ctx, entry.m_Priority, entry.m_Name, entry.m_Uri, true);
+    }
+
+    FreeMountsFile(entries);
+    return dmResource::RESULT_OK;
+}
+
+dmResource::Result SaveMounts(HContext ctx, const char* app_support_path)
+{
+    char path[DMPATH_MAX_PATH];
+    dmPath::Concat(app_support_path, MOUNTS_FILENAME, path, sizeof(path));
+
+    DM_RESOURCE_DBG_LOG(1, "Saving mounts '%s'\n", path);
+
+    DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+
+    dmArray<MountFileEntry> entries;
+
+    uint32_t size = ctx->m_Mounts.Size();
+    for (uint32_t i = 0; i < size; ++i)
+    {
+        ArchiveMount& mount = ctx->m_Mounts[i];
+        if (!mount.m_Persist)
+            continue;
+
+        if (entries.Full())
+            entries.OffsetCapacity(8);
+
+        dmURI::Parts uri;
+        dmResourceProvider::GetUri(mount.m_Archive, &uri);
+
+        char uri_str[1024];
+        if (uri.m_Location[0] == '\0')
+            dmSnPrintf(uri_str, sizeof(uri_str), "%s:%s", uri.m_Scheme, uri.m_Path);
+        else
+            dmSnPrintf(uri_str, sizeof(uri_str), "%s:%s/%s", uri.m_Scheme, uri.m_Location, uri.m_Path);
+
+        MountFileEntry entry;
+        entry.m_Name = strdup(mount.m_Name);
+        entry.m_Uri = strdup(uri_str);
+        entry.m_Priority = mount.m_Priority;
+
+        entries.Push(entry);
+    }
+
+    dmResource::Result result = WriteMountsFile(path, entries);
+    FreeMountsFile(entries);
+
+    return result;
+}
+
+dmResource::Result GetMountByIndex(HContext ctx, uint32_t index, SGetMountResult* mount_info)
+{
+    DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+    if (index >= ctx->m_Mounts.Size())
+        return dmResource::RESULT_INVAL;
+
+    ArchiveMount& mount = ctx->m_Mounts[index];
+    mount_info->m_Name = mount.m_Name;
+    mount_info->m_Archive = mount.m_Archive;
+    mount_info->m_Priority = mount.m_Priority;
+    return dmResource::RESULT_OK;
+}
+
+// Unit tests
+dmResource::Result GetMountByName(HContext ctx, const char* name, SGetMountResult* mount_info)
+{
+    DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+    uint32_t size = ctx->m_Mounts.Size();
+    for (uint32_t i = 0; i < size; ++i)
+    {
+        ArchiveMount& mount = ctx->m_Mounts[i];
+        if (strcmp(mount.m_Name, name) == 0)
+            return GetMountByIndex(ctx, i, mount_info);
+    }
+    return dmResource::RESULT_INVAL;
+}
+
+static dmResource::Result DestroyMounts(HContext ctx)
+{
+    uint32_t size = ctx->m_Mounts.Size();
+    for (uint32_t i = 0; i < size; ++i)
+    {
+        ArchiveMount& mount = ctx->m_Mounts[i];
+        free((void*)mount.m_Name);
+        dmResourceProvider::Unmount(mount.m_Archive);
+    }
+    ctx->m_Mounts.SetSize(0);
+    return dmResource::RESULT_OK;
+}
+
+dmResource::Result GetResourceSize(HContext ctx, dmhash_t path_hash, const char* path, uint32_t* resource_size)
```

Comment:
The resource system calls the GetResourceSize() in order to know preallocate memory for the loading part.
There is certainly a case to be made to not do both GetResourceSize()+ReadResource() since the first step has already found the mount and the mount had already done the lookup.

If we need to, we can perhaps create some GetResource() api which can cater for both cases.